marlowe-spec-test/src/Marlowe/Spec/Core/SemiArbitrary.hs

```diff
@@ -0,0 +1,537 @@
+{-# LANGUAGE BlockArguments    #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+{-# LANGUAGE TupleSections     #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Marlowe.Spec.Core.SemiArbitrary
+  ( arbitraryContractWeighted
+  , arbitraryTransaction
+  , arbitraryValidTransactions
+  , genContext
+  , Context (..)
+  , SemiArbitrary (..)
+  )
+  where
+
+import qualified Arith
+import Control.Monad (replicateM)
+import Data.Function (on)
+import Data.List (nubBy)
+import Data.Map (Map, fromList, toList)
+import qualified Examples.Escrow as Escrow
+import qualified Examples.Swap as Swap
+import Marlowe.Spec.Core.Arbitrary
+  ( arbitraryChoiceName,
+    arbitraryInteger,
+    arbitraryNonnegativeInteger,
+    arbitraryPositiveInteger,
+    arbitraryTimeIntervalAfter,
+    arbitraryTimeIntervalBefore,
+    chooseinteger,
+    genBound,
+    genTransactionError,
+    genValueId,
+  )
+import Marlowe.Spec.Core.Generators (genParty, genToken)
+import Marlowe.Spec.Interpret
+  ( InterpretJsonRequest,
+  )
+import Orderings (Ord (..), max)
+import QuickCheck.GenT
+  ( Gen,
+    GenT,
+    MonadGen (..),
+    frequency,
+    resize,
+    sized,
+    vectorOf,
+  )
+import Semantics
+  ( Payment (..),
+    TransactionOutput (..),
+    TransactionOutputRecord_ext (..),
+    TransactionWarning (..),
+    Transaction_ext (..),
+    computeTransaction,
+    evalValue,
+  )
+import SemanticsTypes
+  ( Action (..),
+    Bound (..),
+    Case (..),
+    ChoiceId (..),
+    Contract (..),
+    Environment_ext (..),
+    Input (..),
+    Observation (..),
+    Party (..),
+    Payee (..),
+    State_ext (..),
+    Token (..),
+    Value (..),
+    ValueId (..),
+    minTime,
+  )
+import Test.QuickCheck (chooseInt, elements)
+import Test.QuickCheck.Arbitrary (Arbitrary (..))
+import qualified Test.QuickCheck.Gen as QC (elements)
+
+-- | Class for arbitrary values with respect to a context.
+class SemiArbitrary a where
```

Comment:
We could also consider a generic type-class that covers both cases, i.e. the `SemiArbitrary` and `Generators` use cases. Something like:

```
-- | Class for arbitrary values with respect to a context.
class SemiArbitrary m a where
 type C m

 -- | Generate an arbitrary value within a context.
 semiArbitrary :: Monad m => C m -> GenT m a

```
This would allow to share a lot of code.